File: 백준/Gold/14621. 나만 안되는 연애/나만 안되는 연애.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Comparator;
import java.util.PriorityQueue;
import java.util.StringTokenizer;

public class Main {

    static int[] parent;

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());

        String[] arr = br.readLine().split(" ");

        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        PriorityQueue<Edge> qu = new PriorityQueue<>(Comparator.comparingInt(a -> a.cost));

        for (int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());

            int from = Integer.parseInt(st.nextToken()) - 1;
            int to = Integer.parseInt(st.nextToken()) - 1;
            int cost = Integer.parseInt(st.nextToken());

            if (arr[from].equals(arr[to])) {
                continue;
            }

            qu.offer(new Edge(from, to, cost));
        }

        int ans = 0;
        int link = 0;

        while (!qu.isEmpty()) {

            Edge e = qu.poll();

            if (find(e.from) != find(e.to)) {
                union(e.from, e.to);
                ans += e.cost;
                link++;
            }
        }

        System.out.println(link == n - 1 ? ans : -1);
    }

    public static int find(int a){
        if (parent[a] == a) {
            return a;
        }

        return parent[a] = find(parent[a]);
    }

    public static void union(int a, int b) {
        a = find(a);
        b = find(b);
        if (a != b) {
            parent[b] = a;
        }
    }

    static class Edge {
        int from, to, cost;

        public Edge(int from, int to, int cost) {
            this.from = from;
            this.to = to;
            this.cost = cost;
        }
    }
}

File: 백준/Gold/14621. 나만 안되는 연애/나만 안되는 연애.py
import sys
from queue import PriorityQueue

input = sys.stdin.readline

n, m = map(int, input().split())
schools = input().split()
parent = [i for i in range(n)]


def find(a):
    if parent[a] == a:
        return a
    parent[a] = find(parent[a])
    return parent[a]


def union(a, b):
    a = find(a)
    b = find(b)
    if a != b:
        parent[b] = a


qu = PriorityQueue()

for _ in range(m):
    start, end, cost = map(int, input().split())

    start -= 1
    end -= 1

    if schools[start] == schools[end]:
        continue

    qu.put((cost, start, end))

ans = 0
link = 0

while not qu.empty():
    cost, start, end = qu.get()

    if find(start) != find(end):
        union(start, end)
        ans += cost
        link += 1

print(ans if link == n - 1 else -1)